python/ray/data/impl/split.py

```diff
@@ -0,0 +1,1104 @@
+        self.bins = {bin_.id: bin_ for bin_ in bins}
+
+    @classmethod
+    def build(cls, bin_ids: List[BinID], capacity: int):
+        """Create a BinSet with uniform capacity.
+        """
+        return cls([Bin(bin_id, capacity) for bin_id in bin_ids])
+
+    def get_bin(self, key: Union[BinID, int]) -> Bin:
+        """Get a bin. The provided key can either be a bin identifier or an
+        integer index; in the latter case, the integer index key will give the
+        corresponding bin according to insertion order.
+        """
+        try:
+            return self.bins[key]
+        except KeyError:
+            return list(iter(self))[key]
+
+    def get_bin_subset(self, bin_ids: List[BinID]) -> "BinSet":
+        """Get a subset of bins.
+
+        Args:
+            bin_ids: The ids of bins that we want in the subset.
+
+        Returns:
+            A BinSet containing only the bins that correspond to the provided
+            bin IDs.
+        """
+        return BinSet([self.get_bin(bin_id) for bin_id in bin_ids])
+
+    def get_bin_with_largest_slack(self) -> Bin:
+        """Get the bin with the most slack.
+        """
+        return max(iter(self), key=lambda bin_: bin_.slack)
+
+    def get_bin_with_best_fit(self, block: BlockData) -> Bin:
+        """Get bin with the best (tightest) fit.
+        """
+        return min(iter(self), key=lambda bin_: bin_.slack - block.num_rows)
+
+    def get_bins_ordered_by_slack(self, descending: bool = True) -> List[Bin]:
+        """Get all available (not full) bins ordered by slack.
+
+        Args:
+            descending: Whether to order the list in descending order.
+
+        Returns:
+            A list of bins in sorted slack order, with full bins removed.
+        """
+        # TODO(Clark): Use a slack-based priority queue for the available bin
+        # set so we don't have to continuously re-sort and recalculate the
+        # max slack.
+        bins = filter(lambda bin_: bin_.slack > 0, iter(self))
+        return sorted(bins, key=lambda bin_: bin_.slack, reverse=descending)
+
+    def is_empty(self):
+        """Whether this bin set is empty.
+        """
+        return len(self.bins) == 0
+
+    def __iter__(self):
+        return iter(self.bins.values())
+
+    def __getitem__(self, key):
+        return self.get_bin(key)
+
+
+def allocate_blocks_to_primary_actors(
+        block_set: BlockSet,
+        actor_bin_set: BinSet,
+        actors_by_node_id: Dict[str, ray.actor.ActorHandle],
+        overallocate: bool = False):
+    """Attempt to allocate full blocks to actors on primary nodes.
+    """
+    primary_actors_by_block = {
+        b: actors_by_node_id[b.preferred_nodes[0]]
+        for b in block_set.get_unallocated_blocks()
+    }
+    _allocate_blocks_to_actors(
+        block_set,
+        actor_bin_set,
+        lambda b: primary_actors_by_block[b],
+        overallocate=overallocate)
+
+
+def allocate_blocks_to_secondary_actors(
```

Comment:
The core logic for primary locations and secondary locations is exactly the same (negligible added complexity), and removing the secondary locations round would only remove a few lines of code, so I'm not sure if removing that round is a good complexity vs. value add tradeoff?

The greatest reduction in complexity would actually be treating all object locations as equal and doing a single locality-aware full-bin packing round, since we then wouldn't need logic picking out the single primary location.

Reply:
yeah that's another option